FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -27,16 +27,53 @@ public final class CoreDataFeedStore: FeedStore {
 		)
 		context = container.newBackgroundContext()
 	}
+}
 
+extension CoreDataFeedStore: FeedStore {
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				guard let cache = try ManagedCache.find(in: context) else {
+					completion(.empty)
+					return
+				}
+				let feed = cache.feed.compactMap { ($0 as? ManagedFeedImage)?.local }
+				completion(.found(feed: feed, timestamp: cache.timestamp))
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				let newCache = try ManagedCache.newUniqueInstance(in: context)
+				newCache.feed = ManagedFeedImage.images(from: feed, in: context)
+				newCache.timestamp = timestamp
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				try ManagedCache.find(in: context).map(context.delete)
+				try context.save()
```

Comment:
In case there is no previously stored cache `context.delete` will not be called, thus there will be no changes in the context and will be no need to save it. Otherwise, Core Data may perform unnecessary work.

Please move `context.save` call inside an `if let ...` block
```suggestion
				if let cache = try ManagedCache.find(in: context) {
					context.delete(cache)
					try context.save()
				}
```

Reply:
Thank you. I fixed it.